enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -0,0 +1,801 @@
+		return err
+	}
+	e.atimes.Delete(xxhash.Sum64(sample.fileMetadataKey))
+	return disk.DeleteFile(context.TODO(), sample.filePath)
+}
+
+func (e *partitionEvictor) resampleK(k int) error {
+	iter := e.iter()
+	defer iter.Close()
+
+	for i := 0; i < k; i++ {
+		entries, err := e.randomSample(iter, sampleN)
+		if err != nil {
+			return err
+		}
+		for _, entry := range entries {
+			e.samplePool = append(e.samplePool, entry)
+		}
+	}
+
+	for _, sample := range e.samplePool {
+		if err := e.refreshAtime(sample); err != nil {
+			continue
+		}
+	}
+
+	if len(e.samplePool) > 0 {
+		sort.Slice(e.samplePool, func(i, j int) bool {
+			return e.samplePool[i].timestamp < e.samplePool[j].timestamp
+		})
+	}
+
+	if len(e.samplePool) > samplePoolSize {
+		e.samplePool = e.samplePool[:samplePoolSize]
+	}
+
+	return nil
+}
+
+func (e *partitionEvictor) evict(count int) error {
+	evicted := 0
+	for evicted < count {
+		lastCount := evicted
+
+		// Resample every time we evict a key
+		if err := e.resampleK(1); err != nil {
+			return err
+		}
+		for i, sample := range e.samplePool {
+			fileMetadataKey, err := constants.FileMetadataKey(sample.fileRecord)
+			if err != nil {
+				return err
+			}
+			_, closer, err := e.reader.Get(fileMetadataKey)
+			if err == pebble.ErrNotFound {
+				continue
+			}
+			closer.Close()
+			if err := e.deleteFile(sample); err != nil {
+				continue
+			}
+			evicted += 1
+			e.samplePool = append(e.samplePool[:i], e.samplePool[i+1:]...)
+			break
+		}
+
+		// If no candidates were evictable in the whole pool, resample
+		// the pool.
+		if lastCount == evicted {
+			e.samplePool = e.samplePool[:0]
+			if err := e.resampleK(samplePoolSize); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (e *partitionEvictor) computeSize() error {
+	iter := e.iter()
+	defer iter.Close()
+
+	casPrefix := []byte(e.part.ID + "/cas/")
+	acPrefix := []byte(e.part.ID + "/ac/")
+	totalCount := int64(0)
+	casCount := int64(0)
+	acCount := int64(0)
+	blobSizeBytes := int64(0)
+	metadataSizeBytes := int64(0)
+	fileMetadata := &rfpb.FileMetadata{}
+
+	for iter.Next() {
+		if err := proto.Unmarshal(iter.Value(), fileMetadata); err != nil {
+			return err
+		}
+		blobSizeBytes += fileMetadata.GetFileRecord().GetDigest().GetSizeBytes()
+		metadataSizeBytes += int64(len(iter.Value()))
+		totalCount += 1
+
+		// identify and count CAS vs AC files.
+		if bytes.Contains(iter.Key(), casPrefix) {
+			casCount += 1
+		} else if bytes.Contains(iter.Key(), acPrefix) {
+			acCount += 1
+		} else {
+			log.Warningf("Unidentified file (not CAS or AC): %q", iter.Key())
+		}
+	}
+
+	e.totalCount = totalCount
+	e.totalSizeBytes = blobSizeBytes // + metadataSizeBytes
+	e.casCount = casCount
+	e.acCount = acCount
+	return nil
+}
+
+func (e *partitionEvictor) ttl() error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	maxAllowedSize := int64(janitorCutoffThreshold * float64(e.part.MaxSizeBytes))
+	for {
+		if err := e.computeSize(); err != nil {
```

Comment:
What's the overhead of recomputing this size every second?

Reply:
higher than I want it to eventually be, but I think it's fine for a mvp.